### DISCUSSION TITLE
Deferrable operators have high CPU usage

### DISCUSSION THREAD

**Discussion by tomsej** [apache/airflow#29430] [Q&A]
Hi, I have been playing with deferrable operators in Airflow, and it seems that they consume a lot of CPU.

**System setup:**
1x Worker, ECS Fargate, cpu=2048, memory=4096
1x Scheduler, ECS Fargate, cpu=2048, memory=4096
1x Triggerer, ECS Fargate, cpu=2048, memory=4096
Elasticache, cache.t4g.micro, engine_version=7.0
It uses a local filesystem for distributing DAGs.

**Airflow setup:**
Docker: apache/airflow:2.5.1-python3.10
No special configuration is provided.

**Custom Defferable Operator:**

```from datetime import timedelta

from airflow.models.baseoperator import BaseOperator
from airflow.triggers.temporal import TimeDeltaTrigger


class HelloOperator(BaseOperator):
    def __init__(self, **kwargs) -> None:
        super().__init__(**kwargs)

    def execute(self, context):
        self.defer(
            trigger=TimeDeltaTrigger(timedelta(seconds=300)),
            method_name="execute_complete",
        )

    def execute_complete(self, context, event):
        pass
```

**Dag:**

```
import pendulum
from airflow.models.dag import DAG
from utils.operator import HelloOperator

with DAG(
    dag_id="hello",
    start_date=pendulum.datetime(
        year=2023, month=1, day=1, hour=0, minute=0, tz="Europe/Prague"
    ),
    schedule="@daily",
    max_active_runs=1,
    max_active_tasks=1,
    catchup=False,
):
    HelloOperator(task_id=f"task-1")
```
If I run the DAG, I see a big peak in the Cloudwatch:
<img width="1250" alt="image" src="https://user-images.githubusercontent.com/6624716/217578676-4345b2f5-a21c-4069-92c0-8a52e8b543e4.png">

This starts to be a problem if you simultaneously run multiple deferred operators, since CPU goes to 100% and some of the deferred tasks fail without any error message in the log:
```
[2023-02-08, 15:04:03 CET] {local_task_job.py:208} INFO - Task exited with return code 0
[2023-02-08, 15:04:03 CET] {taskinstance.py:2578} INFO - 0 downstream tasks scheduled from follow-on schedule check
```

But in the worker, an error appears:
```
Executor reports task instance <TaskInstance: x [queued]> finished (success) although the task says its queued. (Info: None) Was the task killed externally?
```

Can be ralated to https://github.com/apache/airflow/issues/10790

**Comment by notatallshaw-gts** [comment]
I do not see this CPU usage problem but FYI I put all async operators in a pool as separately the worker VMs I use have had memory constraints.

Putting async operators in a pool only affects the worker process spawning and not how many triggerer tasks can run at any given time. So let's say you set the pool size to 5 then if you kick off 100 async operators only 5 will start/stop simultaneously, in my experience you will have all 100 triggers running in less than 30 seconds but without the "shock" to the workers resources.

**Reply by tomsej** [reply]
Thanks @notatallshaw-gts, I understand that it can be solved with pool, worker concurrency or parallelism. I am honestly doing the same as you, but it feels hacky and counter-intuitive. Also, I find it interesting that deferring takes up that much CPU (even that simple one). I am wondering why. That is why I created this as a discussion, not as a bug.

**Comment by tomsej** [comment]
Do you have any ideas or tips on what to focus on? @andrewgodwin (btw sorry for tagging you 😀)

**Reply by andrewgodwin** [reply]
Without a lot more detailed metrics, and a comparison of deferring a task to simply ending it (as both unload the task from the worker and do stuff as part of that process), I do not.

It just seems to be a small CPU spike at deferral time rather than throughout the five minutes that the task is deferred for, so that does not seem particularly bad, either?

**Reply by tomsej** [reply]
Thanks! I will try to simulate it locally. The problem I currently have is that if I run ~6 deferred tasks cpu is immediately skyrocketing and oom killer kills the worker. If I run just one task it is more or less fine. So I will try to dig deeper and will let you know.
